buckets/redis/bucket.go

```diff
@@ -56,32 +53,30 @@ func (a *abstractBucket) Take(ctx context.Context, requested int64, maxWaitTime
 		strconv.FormatInt(requested, 10), strconv.FormatInt(maxWaitTime.Nanoseconds(), 10),
 		maxIdleTimeMillis, a.maxDebtNanos}
 
-	var waitTime time.Duration
-	var err error
-
 	client := a.factory.Client().(*redis.Client)
 	res := a.takeFromRedis(ctx, client, args)
-	switch waitTimeNanos := res.Val().(type) {
+	if err := res.Err(); err != nil {
+		if isRedisClientClosedError(err) {
+			logging.Print("Failed to take token from redis because the client was closed, reconnecting")
+			a.factory.handleConnectionFailure(client)
+		}
+		return 0, true, errors.Wrap(err, "failed to take token from redis bucket")
```

Comment:
Should we add a comment explaining the non-intuitive "successful" return?

Reply:
It's relatively intuitive if you expect the code to fail open. I'm changing it to be a `false` because it has a little less cognitive overhead when glancing at the code since it follows more expected fail closed behaviour.